### PR TITLE
Add transitions with view transitions API

### PR DESCRIPTION
**Pull request by SantosGuillamot**
Coupled with [this other pull request](https://github.com/WordPress/block-interactivity-experiments/pull/241), we added a new experimental directive to add a transition between featured images using the View Transitions API. Here you have a video with the explanation:

https://www.loom.com/share/39631b65aebf42b386f0fee14e0b3f58?sid=fd5f26d5-1105-43ab-87a0-d66467e6ebff

Basically, what I did:

* Support a new property `viewTransitionsAPI` in the `wp-link` directive. If it is set to true, the viewTransitionsAPI is enabled while navigating. This already enables all the transitions that can be achieved only with CSS.
* For more complex uses cases, I created a `data-wp-transition` directive:
    * It creates a global action and global state to be able to create transitions that need a unique id. The transitions are added dynamically with JS. When the action is called, it updates the global state and add the transition to the current element. In the new page, we have a init function that checks if the key defined in the directive matches the global state. If so, it adds the transition dynamically.

There are some aspects to review, although it is already working more or less fine for the movies:

* There seems to be a problem with the backdrop-filter.
* The JS file is sent each time I visit a post.

### CODE REVIEW

File: wpmovies.php [hunk 1]
```diff
@@ -130,14 +130,47 @@ function wpmovies_add_key_to_featured_image( $content ) {
 	while ( $p->next_tag( array( 'tag_name' => 'img' ) ) ) {
 		$src = $p->get_attribute( 'src' );
 		if ( preg_match( '/\/([\w-]+)\.jpg$/', $src, $matches ) ) {
-			$p->set_attribute( 'key', $matches[1] );
+			$p->set_attribute( 'data-key', $matches[1] );
```
Comment (luisherranz) [inline wpmovies.php:133]: Let's change `data-key` to `data-wp-key` and add `if (vnode.props['data-wp-key']) vnode.props.key = vnode.props['data-wp-key'];` to the hooks in the BIE repo.

File: wpmovies.php [hunk 2]
```diff
@@ -130,14 +130,47 @@ function wpmovies_add_key_to_featured_image( $content ) {
...
+		$p->set_attribute( 'data-wp-on--click', 'actions.core.navigation.addTransition' );
+		$p->set_bookmark( 'parent' );
+		$is_link = false;
+		if ( $p->next_tag( 'a' ) ) {
+			$is_link = true;
+		}
+		$p->seek( 'parent' );
+		$data_key = '';
+		if ( $p->next_tag( 'img' ) ) {
+			$data_key = $p->get_attribute( 'data-key' );
```
Comment (luisherranz) [inline wpmovies.php:160]: We need to change this too.

File: lib/view-transitions.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+				// If the key matches, add transition. If not, remove it.
+				if (state.core.navigation.viewTransitionKey === transitionKey) {
+					element.ref.current.style.viewTransitionName =
+						transitionKey;
+				} else {
+					element.ref.current.style.viewTransitionName = '';
+				}
+
+				return evaluate(key, { context: contextValue });
+			});
```
Comment (luisherranz) [inline lib/view-transitions.js:38]: We can do this to avoid unnecessary runs.

```suggestion
			}, [state.core.navigation.viewTransitionKey]);
```

**Review by luisherranz — APPROVED** [review]
Let's make the `data-wp-key` change and this should be ready to go 🙂

File: wpmovies.php (hunk 1 above)
Comment (SantosGuillamot) [reply wpmovies.php]: Changed 🙂

File: lib/view-transitions.js (hunk 3 above)
Comment (SantosGuillamot) [reply lib/view-transitions.js]: I made this change but the transition from movie to an actor stop working as expected. Although I believe it is more related to the way we are adding the `view-transition` names. I'll explore that deeper.

**Comment by SantosGuillamot** [comment]
I made the changes suggested. There is one issue though. It seems that the JS file is being fetched each time I visit a movie. I made a quick video showing what I mean:


https://github.com/WordPress/wp-movies-demo/assets/34552881/d893ee89-ad91-451d-b776-cc4570661b0e

File: lib/view-transitions.js (hunk 3 above)
Comment (luisherranz) [reply lib/view-transitions.js]: Don't do this then. It's not necessary.

**Comment by luisherranz** [comment]
That's because you have the "Disable cache" option enabled, but that option is always off when the devtools are closed.

**Comment by SantosGuillamot** [comment]
> That's because you have the "Disable cache" option enabled, but that option is always off when the devtools are closed.

I'm not sure about this. Right now, fetching the script multiple times throws a console.log error. And I can see that error even if I have the devtools closed. I made a quick video explaining it:

https://www.loom.com/share/eb25392388a749f98d31c4c1844623e4?sid=4ed71037-4a4e-40eb-ad41-862f0f4a1fd4

**Comment by luisherranz** [comment]
Then maybe there's a problem with the way the router is loading the scripts during the client-side navigation?

I'm not familiar with that logic, so @michalczaplinski, could you please take a look? Thanks! 🙂

**Comment by michalczaplinski** [comment]
Sure thing, I'll take a look on Monday!

**Comment by luisherranz** [comment]
@michalczaplinski, you need to use this "coupled" PR:

- https://github.com/WordPress/block-interactivity-experiments/pull/241

**Comment by michalczaplinski** [comment]
@DAreRodz and I looked at this problem, and we found the culprit. The `<script src="src="http://movies.local/wp-content/plugins/wp-movies-demo/lib/view-transitions.js?ver=1.0.0">` node is being re-rendered on the individual movie page because there is another `<script>` being inserted in-between and Preact doesn't know how to reconcile "correctly": 

https://github.com/WordPress/wp-movies-demo/assets/5417266/6efb5ac0-e765-4f2a-9d0d-e94b399c109c

**Comment by SantosGuillamot** [comment]
So, if I understood it correctly, it is not a problem with the movies implementation but with client-side navigation, right? Should we open a new issue for this and discussed it there?

**Comment by michalczaplinski** [comment]
Yes, that's right! @SantosGuillamot I have now updated the video with better-quality audio, btw.

I'll open a new issue in https://github.com/WordPress/block-interactivity-experiments

**Comment by michalczaplinski** [comment]
Opened the issue: https://github.com/WordPress/block-interactivity-experiments/issues/242

**Comment by cbravobernal** [comment]
I'll close this PR for maintenance, we can reopen it in a future.

**Pull request closed without merging**